### PR TITLE
[TF] Add TensorFlow implementation of GRU and make training loop more flexible

### PR DESCRIPTION
**Pull request by floriandonhauser**
This PR adds a TensorFlow implementation for GRU which is somewhat based on the code of @PHOENIXFURY007  and his PR #1597
The code contains both the implementation from scratch and the concise one with keras.

Some changes had to be made to the training loop and the models from the previous chapters:
- When we initialize RNNModelScratch, we now pass in the function get_params which is executed at initialization to create the learnable weights.
- The training loop does not access these weights before the first model invocation. If we access them to early before invocation, keras implementations will return an empty list. The weights for keras are created during the first model invocation (call).
- The weights of the RNN (and later RNN models like GRU or LSTM) are accessed using model.trainable_variables for the implementation from scratch since this makes it compatible with Keras models. **This decision is up for discussion and we could use an if-else statement depending on the type of the model.**
- Parameters are now never passed in as arguments in addition to the model but always taken from the model using model.trainable_variables. This solves the issue of accessing them too early or having wrong values in params.
- The RNNModel now accepts more than two output values from the keras layer. The first output is always Y and the rest are a saved together as the list state. For RNN and GRU, the list state is just the state value. However for keras implementation for LSTM, we get back 3 values! H and C are then combined into state as is the case for the other frameworks. The code, hence, is compatible with more different RNN layers. I also experimented with adding a lambda layer in between, but this causes too much added complexity in my opinion. **However, I am always thankful for other suggestions.**

I already have a finished LSTM implementation in TensorFlow which I will create a PR for, as soon as this PR is merged.

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by mli** [comment]
Job d2l-en/PR-1680/1 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1680/

File: chapter_recurrent-modern/gru.md [hunk 1]
```diff
@@ -242,6 +251,29 @@ def get_params(vocab_size, num_hiddens, device):
...
+```{.python .input}
+#@tab tensorflow
+def get_params(vocab_size, num_hiddens):
+    num_inputs = num_outputs = vocab_size
+    
+    def normal(shape):
+        return d2l.normal(shape=shape,stddev=0.01,mean=0,dtype=tf.float32)
+
+    def three():
+        return tf.Variable(normal((num_inputs, num_hiddens)), dtype=tf.float32),\
```
Comment (terrytangyuan) [inline chapter_recurrent-modern/gru.md:263]: Use parenthesis to surround these and avoid the use of `\`

File: chapter_recurrent-neural-networks/rnn-concise.md [hunk 2]
```diff
@@ -235,11 +235,13 @@ class RNNModel(tf.keras.layers.Layer):
         self.rnn = rnn_layer
         self.vocab_size = vocab_size
         self.dense = tf.keras.layers.Dense(vocab_size)
-    def call(self, inputs, state, _):
+
+    def call(self, inputs, state):
         X = tf.one_hot(tf.transpose(inputs), self.vocab_size)
-        Y,state = self.rnn(X, state)
+        Y, *state = self.rnn(X, state)
```
Comment (terrytangyuan) [inline chapter_recurrent-neural-networks/rnn-concise.md:241]: Perhaps add a comment here on why the wildcard is needed?

**Review by terrytangyuan — CHANGES_REQUESTED** [review]

File: chapter_recurrent-modern/gru.md (hunk 1 above)
Comment (floriandonhauser) [reply chapter_recurrent-modern/gru.md]: Done

File: chapter_recurrent-neural-networks/rnn-concise.md (hunk 2 above)
Comment (floriandonhauser) [reply chapter_recurrent-neural-networks/rnn-concise.md:242]: Done

**Comment by mli** [comment]
Job d2l-en/PR-1680/2 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1680/

**Review by terrytangyuan — APPROVED** [review]

File: chapter_recurrent-neural-networks/rnn-scratch.md [hunk 3]
```diff
@@ -412,16 +411,16 @@ def predict_ch8(prefix, num_preds, net, vocab, device):  #@save
 
 ```{.python .input}
 #@tab tensorflow
-def predict_ch8(prefix, num_preds, net, vocab, params):  #@save
+def predict_ch8(prefix, num_preds, net, vocab): #@save
```
Comment (astonzhang) [inline chapter_recurrent-neural-networks/rnn-scratch.md:414]: two space before an inline comments

File: chapter_recurrent-neural-networks/rnn-scratch.md [hunk 4]
```diff
@@ -732,20 +731,17 @@ def train_ch8(net, train_iter, vocab, lr, num_epochs, device,
 ```{.python .input}
 #@tab tensorflow
 #@save
-def train_ch8(net, train_iter, vocab, num_hiddens, lr, num_epochs, strategy,
-              get_params, use_random_iter=False):
+def train_ch8(net, train_iter, vocab, num_hiddens, lr, num_epochs, strategy, use_random_iter=False):
```
Comment (astonzhang) [inline chapter_recurrent-neural-networks/rnn-scratch.md:734]: max 80 chars per line

**Comment by astonzhang** [comment]
@floriandonhauser Thanks! I just left some minor comments. Besides, do you want to fix https://github.com/d2l-ai/d2l-en/issues/1678 after this?

**Comment by floriandonhauser** [comment]
> @floriandonhauser Thanks! I just left some minor comments. Besides, do you want to fix #1678 after this?

Thank you for the feedback.
I will create a PR for issue #1678 after I am done with more RNN chapters. I'm currently working on LSTM and will create a PR for it very soon™ .

**Pull request merged**